types/react-dom/test/canary-tests.tsx

```diff
@@ -128,7 +128,27 @@ function formTest() {
             return state + 1;
         }
 
-        const [state, dispatch] = useFormState(action, 1);
+        const [
+            // $ExpectType number
+            state,
+            dispatch,
+        ] = useFormState(action, 1);
+
+        function actionExpectingPromiseState(state: Promise<number>) {
+            return state.then(s => s + 1);
+        }
+
+        useFormState(
+            // @ts-expect-error
+            actionExpectingPromiseState,
+            Promise.resolve(1),
+        );
+        useFormState(
+            action,
+            // @ts-expect-error
+            Promise.resolve(1),
```

Comment:
Omitting the `Awaited` from the initial state does not give me the expected error in [this playground](https://www.typescriptlang.org/play?#code/PTAEFsE9QBwJwPYwQZwIYBsBQATApgMYZpx6gBmArgHYEAuAlgtaJSngGIJzgDKdaOngCMAHl4AaUAAUAfAAosoUGnpNqALlDz4eAG79BeLQEEA7mgZCc42VJhpIGBGhxbpASlABeWaF6gAD4yiOAM7LYSSqAM1FYMmIZCphZWeDa8dtEweDyYsQDWAPxaKHRwsQDmUR5aANrmltaR2g5OLm4yXr6geggMOAC6ANxYWGyc3HwCQsLyOqQGM3jdfrpLRlIADB6jE1w8SSLyaCiQtK2LR6uwV8vbu1ggoAACdCgAtHgAHjn0X3BEHBxuwDtMjHNTucCJd9NcfGs7psQggwuwAHSkFAIDB6PDyHaPJ5gAAWdDoMBQGhAlSsJMoACN0QRUcAACJ4cixNIYSAAFUgORw7M53KEvIFQuAMEoGAwwAAbAB2ABMCoVuEIxFIFBoamYrFBUyOKvEUjkimUqkYzC0CzhyxSTXSLTazlc7huAWC0lC4TwkWiYoSGCOWkk2Vy4Hy1GKpXKVRq9UkrUc7s6ngRvX6Q1GIMmh2WKvm63hPVL91AOz2RsLRmLUIu9o2QhuFeR1eJr3eX1+hDoAKB+bBJpOZyb7dbWcneHNfoxWJxeIJHkeQA)? What's the difference to the changes proposed in this PR? Are the tests running with different compiler settings? Or am I overlooking something?

Reply:
Yours is also missing the overload. I don't think tests here run with different options. But the diff is too big between my version and yours to say for sure what's happening.

Reply:
I've found the difference in this [playground](https://www.typescriptlang.org/play?#code/PTAEFsE9QBwJwPYwQZwIYBsBQATApgMYZpx6gBmArgHYEAuAlgtaJSngGIJzgDKdaOngCMAHn6C8APgAUWUAtBp6TagC5QMlAKEaAggHc0DITnE7pASlABeKaAlDQAH1AAFROAbtzkqQBp5RQZqEwZMRzx9IxM8M0iAoIUYPB5MEIBrAH4NbTgQgHNAyw0AbW1JaONTXyEA0BxvGEECAAsNGWs7UAA3BAYcAF0AblxCYlIKGhVmVnYuHkixSP93NEgMBDQcWSSlGfVNCt1QQ2q42ulV5o2tnA03dc3trvtIl3dPbzxLxMVQEJhCIWKqxeIWP6KFJpDCZHKgPKFYplY5RU4xGoJVaNFDNOhtDo3Z73Na3F62ex9AYjLBYECgVp0OgwFBqEAFEytSgAIwAdAQEOBgAARPDkQFCDCQAAqkBSOBFYoleClsvlwBglAwGGAADYAOwAJl1urGRBIZCotEYszYnG4fAsht+cn+yhthy0IPR53BfleDgsHw8gu+v0C-wl4QwkQ0Kz20PA6Wo2VydHy1CKWBKoHK3pWDSaLXamgDVKGo3w5smVoOc3tiydl1WjzJO1dindqg6qNBmIh1yedweQ-J3XerhDXh8WL2UeBlUDkgjUNSSdhKfhiMzyNzvaXQmxRfxJZkROHpOJZf6FdpaBQkFoU2tqn2Hq9i+olHA3NS1gA3kEpB0JQcAsKioAANSgMIowAL60naCyOpIwgyGepA9JEAbwHgWFBtBwirAADJYoxIQ6SwyPej4EJouH4ZIOGYe8hEkWRdJgAAAnQKAALR4AAHik9ACXAiBwFgFGNqh1EPk+GF4dhFKwCxBEwS2XzsLypAoAgGA9HgMikRx9I8fxQkiXQYkSVJ8yURYaFdswmmhtpun6YZxmWBxdkNihQiGuhDHKd0IXqURoCkeR9kyYFcm0fRalMSp4WSFBGlRaZ3G8QJwmENZqS2dJAV4EFNEKWlQjMUpEWudOeA6XgekGUZJmjGZuWWQVNncH5yGROVBz1d8TUtV57VYEAA). It does not yield the expected error when using a closure where the action param type is inferred instead of explicitly being annotated:

```ts
useFormState(async (prevState) => prevState + 1, Promise.resolve(0));
```